core/sr-arithmetic/src/biguint.rs

```diff
@@ -0,0 +1,825 @@
+				// Double = u16, think of 255 * 256 + 255 which is just u16::max_value().
+				let dividend =
+					Double::from(self_norm.get(j + n))
+						* B
+						+ Double::from(self_norm.get(j + n - 1));
+				let divisor = other_norm.get(n - 1);
+				div_single(dividend, divisor)
+			};
+
+			// D3.1 test qhat
+			// replace qhat and rhat with RefCells. This helps share state with the closure
+			let qhat = RefCell::new(qhat);
+			let rhat = RefCell::new(Double::from(rhat));
+
+			let test = || {
+				// decrease qhat if it is bigger than the base (B)
+				let qhat_local = *qhat.borrow();
+				let rhat_local = *rhat.borrow();
+				let predicate_1 = qhat_local >= B;
+				let predicate_2 = {
+					let lhs = qhat_local * Double::from(other_norm.get(n - 2));
+					let rhs = B * rhat_local + Double::from(self_norm.get(j + n - 2));
+					lhs > rhs
+				};
+				if predicate_1 || predicate_2 {
+					*qhat.borrow_mut() -= 1;
+					*rhat.borrow_mut() += Double::from(other_norm.get(n - 1));
+					true
+				} else {
+					false
+				}
+			};
+
+			test();
+			while (*rhat.borrow() as Double) < B {
+				if !test() { break; }
+			}
+
+			let qhat = qhat.into_inner();
+			// we don't need rhat anymore. just let it go out of scope when it does.
+
+			// step D4
+			let lhs = Self { digits: (j..=j+n).rev().map(|d| self_norm.get(d)).collect() };
+			let rhs = other_norm.clone().mul(&Self::from(qhat));
+
+			let maybe_sub = lhs.sub(&rhs);
+			let mut negative = false;
+			let sub = match maybe_sub {
+				Ok(t) => t,
+				Err(t) => { negative = true; t }
+			};
+			(j..=j+n).for_each(|d| { self_norm.set(d, sub.get(d - j)); });
+
+			// step D5
+			// PROOF: the `test()` specifically decreases qhat until it is below `B`. conversion
+			// is safe.
+			q.set(j, qhat as Single);
+
+			// step D6: add back if negative happened.
+			if negative {
+				q.set(j, q.get(j) - 1);
+				let u = Self { digits: (j..=j+n).rev().map(|d| self_norm.get(d)).collect() };
+				let r = other_norm.clone().add(&u);
+				(j..=j+n).rev().for_each(|d| { self_norm.set(d, r.get(d - j)); })
+			}
+		}
+
+		// if requested, calculate remainder.
+		if rem {
+			// undo the normalization.
+			if normalizer_bits > 0 {
+				let s = SHIFT as u32;
+				let nb = normalizer_bits;
+				for d in 0..n-1 {
+					let v = self_norm.get(d) >> nb
+						| self_norm.get(d + 1).overflowing_shl(s - nb).0;
+					r.set(d, v);
+				}
+				r.set(n - 1, self_norm.get(n - 1) >> normalizer_bits);
+			} else {
+				r = self_norm;
+			}
+		}
+
+		Some((q, r))
+	}
+}
+
+#[cfg(feature = "std")]
+impl rstd::fmt::Debug for BigUint {
+	fn fmt(&self, f: &mut rstd::fmt::Formatter<'_>) -> rstd::fmt::Result {
+		write!(
+			f,
+			"BigUint {{ {:?} ({:?})}}",
+			self.digits,
+			u128::try_from(self.clone()).unwrap_or_else(|_| 0),
+		)
+	}
+}
+
+impl PartialEq for BigUint {
+	fn eq(&self, other: &Self) -> bool {
+		// sadly, we have to reallocate here as strip mutably uses self.
```

Comment:
Hmm I seem to have forgotten to do this: 

I think this can just use `self.cmp(other) == Ordering::Equal`. Or a similar implementation, we can avoid re-allocation here.

Reply:
but we can do in following PR